src/v/raft/state_machine_manager.cc

```diff
@@ -239,6 +245,82 @@ ss::future<> state_machine_manager::stop() {
     co_await std::move(gate_f);
 }
 
+ss::future<> state_machine_manager::apply_initial_recovery_policy() {
+    auto snapshot = co_await read_initial_recovery_snapshot();
+    if (!snapshot) {
+        /**
+         * If snapshot is not available create one, it will be populated during
+         * initial recovery
+         */
+        vlog(_log.debug, "no initial recovery snapshot found");
+        snapshot.emplace(initial_recovery_snapshot{});
+    }
+    vlog(_log.debug, "Starting with initial recovery snapshot: {}", *snapshot);
+    for (auto& [name, entry] : _machines) {
+        auto it = snapshot->initial_recovery_next_offsets.find(name);
+        if (it != snapshot->initial_recovery_next_offsets.end()) {
+            vlog(
+              _log.debug,
+              "skipping initial recovery for {} state machine as it was "
+              "recovered from snapshot",
+              name);
+            entry->stm->set_next(std::max(it->second, entry->stm->next()));
+            continue;
+        }
+        // Stm needs initial recovery
```

Comment:
Even if it has a local snapshot and it's the first time we are running the new Redpanda version?